### PR TITLE
Simple HTTP cache control middleware

### PR DESCRIPTION
**Pull request by dheavy**
I'd be happy to discuss with you the relevance and implementation of this feature in this pull request.

This middleware provides a simple abstraction to set the [`max-age` value of the HTTP `Cache-Control`](https://en.wikipedia.org/wiki/Web_cache#Cache_control) header in a fluent way. One could rely on `express` usage of `Etag` mechanism by default and not even pay attention — that's fine. But in cases where a finer-grained, per-resource HTTP caching strategy is involved, it could come handy. 
### Changes

`cache` is now labelled `ramCache` and API changes accordingly.
### API

``` javascript
var router = dolman.router([
  // Use `httpCache`. Setting a string value equals setting `Cache-Control` header manually.
  {
    url: '/foo',
    httpCache: 'public, max-age=3600, s-maxage=3600' 
  }

  // Setting using an object, where key is the time unit and value is the corresponding value.
  // Accept `seconds` up to `weeks`. Doesn't combine values. If several values are given,
  // only the last one is taken into account.
  {
    url: '/bar',
    httpCache: {minutes: 30}
  },
  {
    url: '/baz',
    ramCache: 'foobarbaz', // still works, mind the name change!
    httpCache: {
      hours: 1,
      minutes: 30,
      seconds: 25
    } // => useless - ultimately set to 25 seconds
  }
]);
```

### CODE REVIEW

**Comment by Yomguithereal** [comment]
This would force the browser to check for a `304` answer and uses its cache, is that it?

**Comment by dheavy** [comment]
**TL;DR** 
In essence, yes.

---

I wouldn't say "force". It's still up to the client to observe this header. 

I'm no expert but from what I know, the `Cache-Control` header will "trigger" caching in the browser
(and for capable proxies along the way unless prevented by `Cache-Control: private`), AND
provide simple information as to how long before the cache could normally go stale. 
You'd also send another tag along enabling estimation of , e.g. `Last-Modified` or `Etag`... which is what `connect`/`express` already does. 

`Cache-Control` would most likely be the knob you'd play around the most if you wanted to get fancy-but-not-overkill with HTTP cache control on your resources (e.g. forcing cache busting on **all** requests on a specific resource: `Cache-Control: no-cache, no-store, must-revalidate`).

So this feature is merely an abstraction for conveniently setting this header in the response.

**Comment by Yomguithereal** [comment]
Fair enough. However, I don't feel that the `ramCache` key is future-proof since we envisioned to add a possibility to plug store into it etc. I guess that let it remain `cache` would be better for the time being.

File: src/middlewares.js [hunk 1]
```diff
@@ -42,6 +42,46 @@ function cache(store, params) {
...
+  return function(req, res, next) {
+    var header = null;
+
+    if (typeof params === 'string')
+      header = params;
+
+    // If params is an object, use special properties
+    // defined by API to set `max-age`.
+    if (params !== null &&
+       (typeof params === 'function') || typeof params === 'object') {
```
Comment (Yomguithereal) [inline src/middlewares.js]: Why the first `typeof` between parentheses and not the second one?

File: src/middlewares.js [hunk 2]
```diff
@@ -42,6 +42,46 @@ function cache(store, params) {
...
+
+      var factor = 0,
```
Comment (Yomguithereal) [inline src/middlewares.js]: The part below seems clunky. I am sure we can find better.

File: src/middlewares.js [hunk 3]
```diff
@@ -42,6 +42,39 @@ function cache(store, params) {
...
+function httpCache(params) {
+  return function(req, res, next) {
+    var header = null;
+
+    if (typeof params === 'string')
+      header = params;
+
+    // If params is an object, use special properties
+    // defined by API to set `max-age`.
+    if (params !== null && typeof params === 'function' || typeof params === 'object') {
```
Comment (dheavy) [inline src/middlewares.js]: @Yomguithereal thoughts on implementation?

Comment (Yomguithereal) [inline src/middlewares.js]: That seems better. However weren't there errors before in case of wrong inputs?

File: src/middlewares.js [hunk 4]
```diff
@@ -42,6 +42,42 @@ function cache(store, params) {
...
+          header = 'private, max-age=' + durations[key] * params[key];
+          return header;
+        }
+      });
+    }
+
+    if (header) {
+      res.set('Cache-Control', header);
+    } else {
+      return next(new Error('Wrong parameter given for HTTP cache control'));
```
Comment (dheavy) [inline src/middlewares.js]: @Yomguithereal Apart from implementation discussion... do you have a way to silence the exception thrown in console when running tests?

Comment (Yomguithereal) [inline src/middlewares.js]: assert.throws
